cli/pkg/kctrl/cmd/package/installed/create_or_update.go

```diff
@@ -957,6 +1006,78 @@ func (o *CreateOrUpdateOptions) showVersions(client pkgclient.Interface) error {
 	return nil
 }
 
+const (
+	yttOverlayPrefix = "ext.packaging.carvel.dev/ytt-paths-from-secret-name"
+)
+
+func (o *CreateOrUpdateOptions) createOrUpdateYttOverlaySecrets(pkgi *kcpkgv1alpha1.PackageInstall, client kubernetes.Interface) (*corev1.Secret, error) {
+	o.statusUI.PrintMessage("Creating overlay secrets")
+	if len(o.YttOverlayFlags.yttOverlayFiles) == 0 {
+		return nil, nil
+	}
+
+	if pkgi != nil {
+		for annotation := range pkgi.Annotations {
+			// Ensure that kctrl does not clobber existing overlays
+			if strings.HasPrefix(annotation, yttOverlayPrefix) && annotation != yttOverlayAnnotation {
+				return nil, fmt.Errorf("Package install has manually supplied overlays")
+			}
+		}
+	}
+
+	secret, err := NewYttOverlays(o.YttOverlayFlags.yttOverlayFiles, o.Name, o.NamespaceFlags.Name).OverlaysSecret()
+	if err != nil {
+		return nil, fmt.Errorf("Structuring overlays secret: %s", err.Error())
+	}
+
+	var createdOrUpdatedSecret *corev1.Secret
+	createdOrUpdatedSecret, err = client.CoreV1().Secrets(o.NamespaceFlags.Name).Create(context.Background(), secret, metav1.CreateOptions{})
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			o.statusUI.PrintMessagef("Updating existing overlay secret '%s' in namespace '%s'", secret.Name, secret.Name)
+			return nil, fmt.Errorf("Creating overlays secret: %s", err.Error())
+		}
+		createdOrUpdatedSecret, err = client.CoreV1().Secrets(o.NamespaceFlags.Name).Update(context.Background(), secret, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("Updating overlays secret: %s", err.Error())
+		}
+	}
+
+	return createdOrUpdatedSecret, nil
+}
+
+func (o *CreateOrUpdateOptions) dropYttOverlaySecrets(pkgi *kcpkgv1alpha1.PackageInstall, client kubernetes.Interface) error {
+	o.statusUI.PrintMessage("Dropping overlay secrets")
+	overlaySecretName, hasOverlay := pkgi.Annotations[yttOverlayAnnotation]
+	if !hasOverlay {
+		return nil
+	}
+
+	overlaySecret, err := client.CoreV1().Secrets(o.NamespaceFlags.Name).Get(context.Background(), overlaySecretName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			o.statusUI.PrintMessagef("Overlay secret '%s' not found in namespace '%s'", overlaySecretName, o.NamespaceFlags.Name)
+			return nil
+		}
+		return fmt.Errorf("Getting overlay secret: %s", err.Error())
+	}
+	annValue, found := overlaySecret.Annotations[KctrlPkgAnnotation]
+	if !found || annValue != o.createdAnnotations.PackageAnnValue() {
+		return fmt.Errorf("Overlay secret was not created by kctrl")
+	}
+
+	err = client.CoreV1().Secrets(o.NamespaceFlags.Name).Delete(context.Background(), overlaySecretName, metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("Deleting overlays secret: %s", err)
+	}
+
+	return nil
+}
+
+func (o *CreateOrUpdateOptions) removeYttOverlaysAnnotation(pkgi *kcpkgv1alpha1.PackageInstall) {
+	delete(pkgi.Annotations, yttOverlayAnnotation)
+}
```

Comment:
Maybe we can do this inline instead of separate function?

Reply:
I did that initially but I thought this helps ad context

Reply:
It just felt weird to have a one liner function. I don't have any strong opinions though.